p2p/net/swarm/swarm_transport.go

```diff
@@ -19,25 +19,23 @@ func (s *Swarm) TransportForDialing(a ma.Multiaddr) transport.Transport {
 
 	s.transports.RLock()
 	defer s.transports.RUnlock()
+
 	if len(s.transports.m) == 0 {
 		// make sure we're not just shutting down.
 		if s.transports.m != nil {
 			log.Error("you have no transports configured")
 		}
 		return nil
 	}
-
-	for _, p := range protocols {
-		transport, ok := s.transports.m[p.Code]
-		if !ok {
-			continue
-		}
-		if transport.Proxy() {
-			return transport
+	if isRelayAddr(a) {
+		return s.transports.m[ma.P_CIRCUIT]
+	}
+	for _, t := range s.transports.m {
```

Comment:
I think in general this needs to go the other way. We should be interpreting this right to left.

Reply:
I don't understand this comment. We're iterating over a `map[int]transport.Transport` here. Inverting the order of iteration shouldn't have any effect, since map iterations are (pseudo-) randomized anyway.

Reply:
I misunderstood. I thought we were iterating of the multiaddr components. I see now that we're iterating over all registered transports.

I think this isn't ideal because:
* we have to make sure every registered transport doesn't intersect with any other transport. It would be tricky behavior if two transports return `CanDial -> true` for the same `a` since we would randomly pick one of the two. I think simply a guaranteed ordering here would make it better.
* This means the transport owns the interpretation of the full multiaddr. This doesn't compose well. Imagine we have a `percentencode` component and a `lenprefixencode` component, we don't want every transport to have to know how to read the encoded data from the multiaddr. Instead they should be given the decoded data from something else and deal with their part of the multiaddr. This relates to https://github.com/multiformats/multiaddr/pull/140.


I'll draft a patch of what I'm thinking here and how it relates to https://github.com/multiformats/multiaddr/pull/140

Reply:
Here's concretely what I mean: https://github.com/libp2p/go-libp2p/compare/fix-dial-transport-selection...marco/layered-transports?expand=1#diff-1d9e588fd7c4fedee40f7f985be2346b02ba540389251f4603cbc13e8f8fc371R94

Notice this gives us the benefit of removing the "extract certhash" logic out of webtransport completely both in the `Dial` and the `CanDial` method (I didn't actually do this refactor, since this is a demo).